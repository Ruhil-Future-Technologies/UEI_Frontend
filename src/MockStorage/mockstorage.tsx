/* eslint-disable @typescript-eslint/no-explicit-any */
import { Box } from '../Pages/AcademicHistory/AcademicHistory';

export const mockStorage: any = {
  menulist1: JSON.stringify([
    {
      created_at: 'Tue, 11 Jun 2024 12:15:27 GMT',
      form_data: {
        form_url: '/main/SuperAdmin',
        is_save: true,
        is_search: true,
        is_update: true,
      },
      id: 23,
      is_active: 1,
      is_deleted: false,
      menu_name: 'Menu',
      priority: 1,
      submenus: [
        {
          created_at: 'Mon, 24 Jun 2024 09:58:08 GMT',
          id: 56,
          is_active: 1,
          is_deleted: false,
          menu_master_id: 23,
          menu_master_name: 'User Authorization',
          menu_name: 'Menu',
          priority: 2,
          updated_at: 'Mon, 24 Jun 2024 13:37:35 GMT',
        },
      ],
      updated_at: 'Fri, 14 Jun 2024 07:40:04 GMT',
    },
    {
      created_at: 'Tue, 11 Jun 2024 12:15:27 GMT',
      form_data: {
        form_url: '/main/SuperAdmin',
        is_save: true,
        is_search: true,
        is_update: true,
      },
      id: 24,
      is_active: 1,
      is_deleted: false,
      menu_name: 'Submenu',
      priority: 1,
      submenus: [
        {
          created_at: 'Mon, 24 Jun 2024 09:58:08 GMT',
          id: 57,
          is_active: 1,
          is_deleted: false,
          menu_master_id: 24,
          menu_master_name: 'User Authorization',
          menu_name: 'Submenu',
          priority: 2,
          updated_at: 'Mon, 24 Jun 2024 13:37:35 GMT',
        },
      ],
      updated_at: 'Fri, 14 Jun 2024 07:40:04 GMT',
    },
  ]),
};

export const contextValue = {
  namepro: { id: 73, user_type: 'admin', userid: 'ashish7080@gmail.com' },
  setNamepro: jest.fn(),
  logoutpro: jest.fn(),
  setProImage: jest.fn(),
  proImage: '',
  setProPercentage: jest.fn(),
  ProPercentage: '',
  setNamecolor: jest.fn(),
  namecolor: '#ffffff',
  activeForm: 0, // Assign a number value, e.g., 0
  setActiveForm: jest.fn(), // Mock function remains unchanged
};

export const mockBoxes: Box[] = [
  {
    id: 334,
    institute_type: 'college',
    board: '',
    state_for_stateboard: '',
    institute_id: '95',
    course_id: 87,
    learning_style: 'online',
    class_id: '24',
    year: '2015-12-31T18:30:00.000Z',
    stream: '',
    university_id: '',
    sem_id: 1,
  },
];

export const mockClassValue = {
  data: [
    {
      class_name: 'class_12',
      created_at: 'Thu, 12 Dec 2024 11:43:04 GMT',
      created_by: 'afasfsafsaf',
      id: '5500819c-09e5-4543-8cf2-8f0d5cf6ffc3',
      is_active: true,
      is_deleted: false,
      updated_at: 'Thu, 12 Dec 2024 11:43:04 GMT',
      updated_by: 'Admin',
    },
  ],
  status: 200,
  message: 'Class Found Successfully',
};

export const mockbasicinfoPayload = {
  payload: [
    {
      student_login_id: '123',
      first_name: 'John',
      last_name: 'Patel',
      gender: 'male',
      dob: '01/01/2024',
      father_name: 'father',
      mother_name: 'mother',
      guardian_name: 'guardian',
      pic_path: 'pic_path',
      aim: 'aim',
    },
  ],
};
