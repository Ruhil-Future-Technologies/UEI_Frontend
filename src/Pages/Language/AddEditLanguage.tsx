/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { useContext, useEffect, useRef, useState } from "react";
import "../Language/Language.scss";
import TextField from "@mui/material/TextField";
import TextareaAutosize from "react-textarea-autosize";
import useApi from "../../hooks/useAPI";
import { useLocation, useNavigate, useParams } from "react-router-dom";
import { QUERY_KEYS_LANGUAGE } from "../../utils/const";
import { Grid, InputLabel, SelectChangeEvent, Typography } from "@mui/material";
import { toast } from "react-toastify";
import { Field, Form, Formik, FormikHelpers } from "formik";
import * as Yup from "yup";
import {
  LanguageRep0oDTO,
  MenuListinter,
} from "../../Components/Table/columns";
import { dataaccess, inputfield, inputfieldtext } from "../../utils/helpers";
import NameContext from "../Context/NameContext";
interface ILanguageForm {
  language_name: string;
  // description:string
}
const AddEditLanguage = () => {
  const context = useContext(NameContext);
  const { namecolor }: any = context;
  const LanguageAddURL = QUERY_KEYS_LANGUAGE.LANGUAGE_ADD;
  const LanguageEditURL = QUERY_KEYS_LANGUAGE.LANGUAGE_EDIT;
  const LanguageURL = QUERY_KEYS_LANGUAGE.GET_LANGUAGE;
  const { getData, postData, putData } = useApi();
  const navigate = useNavigate();
  const { id } = useParams();
  const formRef = useRef() as any;
  const LanguageNamePattern = /^[a-zA-Z\s]*$/;

  const location = useLocation();
  const Menulist: any = localStorage.getItem("menulist1");
  const pathSegments = location.pathname.split("/").filter(Boolean) || [];
  const lastSegment = id
    ? pathSegments[pathSegments.length - 3]?.toLowerCase() || ""
    : pathSegments[pathSegments.length - 2]?.toLowerCase() || "";
  const [filteredData, setFilteredData] = useState<MenuListinter | any>([]);
  const initialState = {
    language_name: "",
  };
  const [language, setLanguage] = useState(initialState);
  const [dataLanguage, setDataLanguage] = useState<LanguageRep0oDTO[]>([]);

  useEffect(() => {
    setFilteredData(
      dataaccess(Menulist, lastSegment, { urlcheck: "" }, { datatest: "" })
    );
  }, [Menulist]);

  if (
    (id && !filteredData?.form_data?.is_update) ||
    (!id && !filteredData?.form_data?.is_save)
  ) {
    navigate("/main/Language");
  }

  useEffect(() => {
    const fetchData = async () => {
      if (id) {
        try {
          // Make sure getData is a valid function returning a promise
          if (typeof getData === "function") {
            const data = await getData(`${LanguageEditURL}/${id}`);
            setLanguage(data?.data || initialState);
          } else {
            throw new Error("getData is not a function");
          }
        } catch (e: any) {
          toast.error(e?.message || "An error occurred while fetching data", {
            hideProgressBar: true,
            theme: "colored",
          });
        }
      }
    };

    fetchData();
  }, [id, LanguageEditURL, getData]);
  const callAPILanguage = async () => {
    getData(`${LanguageURL}`)
      .then((data: any) => {
        if (data?.data) {
          setDataLanguage(data?.data);
        }
      })
      .catch((e) => {
        if (e?.response?.status === 401) {
          // navigate("/")
        }
        // toast.error(e?.message, {
        //     hideProgressBar: true,
        //     theme: "colored",
        // });
      });
  };
  useEffect(() => {
    callAPILanguage();
  }, []);

  const handleChange = async (
    e:
      | React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
      | SelectChangeEvent<string>,
    fieldName: string
  ) => {
    setLanguage((prevMenu) => {
      return {
        ...prevMenu,
        [e.target.name]: e.target.value,
      };
    });

    formRef?.current?.setFieldValue(fieldName, e.target.value);
    if (
      formRef?.current?.errors?.[fieldName as keyof ILanguageForm] !== undefined
    ) {
      formRef?.current?.setFieldError(
        fieldName,
        formRef?.current?.errors?.[fieldName as keyof ILanguageForm]
      );
      formRef?.current?.setFieldTouched(fieldName, true);
    }
  };

  const handleSubmit = async (
    languageData: ILanguageForm,
    { resetForm }: FormikHelpers<ILanguageForm>
  ) => {
    // console.log("test submit", languageData)
    if (id) {
      putData(`${LanguageEditURL}/${id}`, languageData)
        .then((data) => {
          if (data.status === 200) {
            navigate("/main/Language");
            toast.success(data.message, {
              hideProgressBar: true,
              theme: "colored",
            });
          }
        })
        .catch((e) => {
          toast.error(e?.message, {
            hideProgressBar: true,
            theme: "colored",
          });
        });
    } else {
      postData(LanguageAddURL, languageData)
        .then((data) => {
          if (data.status === 200) {
            // navigate('/main/Language');
            callAPILanguage();
            toast.success(data.message, {
              hideProgressBar: true,
              theme: "colored",
            });
            resetForm({ values: initialState });
            setLanguage(initialState);
          }
        })
        .catch((e) => {
          toast.error(e?.message, {
            hideProgressBar: true,
            theme: "colored",
          });
        });
    }
  };
  let languageSchema;
  {
    if (id) {
      languageSchema = Yup.object().shape({
        language_name: Yup.string()
          .required("Please enter Language name")
          .test(
            "not-whitespace",
            "Please enter a valid Language name; whitespace is not allowed.",
            (value: any) => value && value?.trim().length > 0
          )
          .matches(
            LanguageNamePattern,
            "Please enter a valid Language name only characters allowed."
          )
          .test("unique", "Language name already exists", function (value) {
            if (!value) return true;

            // Check if the value matches the current institute name
            if (value.toLowerCase() === language?.language_name.toLowerCase()) {
              return true;
            }

            // Check for uniqueness against dataInstitute
            const exists = dataLanguage?.some(
              (inst) =>
                inst?.language_name &&
                inst?.language_name.toLowerCase() === value?.toLowerCase()
            );

            return !exists;
          }),
        // description: Yup.string()
        // .required("Please enter description ")
      });
    } else {
      languageSchema = Yup.object().shape({
        language_name: Yup.string()
          .required("Please enter Language name")
          .test(
            "not-whitespace",
            "Please enter a valid Language name; whitespace is not allowed.",
            (value: any) => value && value?.trim().length > 0
          )
          .matches(
            LanguageNamePattern,
            "Please enter a valid Language name only characters allowed."
          )
          .test("unique", "Language name already exists", (value) => {
            if (!value) return true;
            const exists = dataLanguage?.some(
              (inst) =>
                inst?.language_name &&
                inst?.language_name?.toLowerCase() === value?.toLowerCase()
            );
            return !exists;
          }),
        // description: Yup.string()
        // .required("Please enter description ")
      });
    }
  }

  return (
    <div className="main-wrapper">
      <div className="main-content">
        <div className="card p-lg-3">
          <div className="card-body">
            <Typography variant="h6" className="mb-3">
              {id ? (
                <div className="main_title">Edit Language</div>
              ) : (
                <div className="main_title">Add Language</div>
              )}
            </Typography>
            <Formik
              onSubmit={(formData, formikHelpers) =>
                handleSubmit(formData, formikHelpers)
              }
              initialValues={{
                language_name: language?.language_name,
              }}
              enableReinitialize
              validationSchema={languageSchema}
              innerRef={formRef}
            >
              {({ errors, values, touched }: any) => (
                <Form>
                  <div className="row">
                    <div className="col-md-4">
                      <div className="form_field_wrapper">
                        <Field
                          inputProps={{ "data-testid": "language_name" }}
                          component={TextField}
                          type="text"
                          label="Language Name *"
                          name="language_name"
                          value={values.language_name}
                          onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
                            handleChange(e, "language_name")
                          }
                        />
                        {errors.language_name && touched.language_name ? (
                          <p style={{ color: "red" }} color="error">
                            {errors.language_name}
                          </p>
                        ) : (
                          <></>
                        )}
                      </div>
                    </div>
                  </div>
                  <div className="row">
                    <div className="col-md-4 mt-2">
                      <div className="col">
                        <Grid item xs={12}>
                          <Typography
                            variant="h6"
                            sx={{ color: inputfieldtext(namecolor) }}
                          >
                            Upload a Photo
                          </Typography>
                        </Grid>
                        <Grid item xs={12}>
                          <input
                            data-testid="language_file"
                            type="file"
                            accept="image/*"
                            id="file-upload"
                            style={{ color: inputfieldtext(namecolor) }}
                          />
                        </Grid>
                      </div>
                    </div>
                  </div>
                  <div className="row mt-4">
                    <div className="col-md-4">
                      <InputLabel
                        className="text-secondary"
                        sx={{ color: inputfieldtext(namecolor) }}
                      >
                        Description
                      </InputLabel>
                      <TextareaAutosize
                        data-testid="language_description"
                        aria-label="empty textarea"
                        minRows={5}
                        style={{
                          width: "100%",
                          fontSize: "1rem",
                          backgroundColor: inputfield(namecolor),
                          color: inputfieldtext(namecolor),
                        }}
                        placeholder="Enter your text here..."
                        name="description"
                        value={values.description}
                        onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) =>
                          handleChange(e, "description")
                        }
                      />
                    </div>
                  </div>
                  <div className="mt-3">
                    <button
                      data-testid="submitBtn"
                      type="submit"
                      className="btn btn-primary mainbutton"
                    >
                      {id ? "Update" : "Save"}
                    </button>
                  </div>
                </Form>
              )}
            </Formik>
          </div>
        </div>
      </div>
    </div>
  );
};

export default AddEditLanguage;
