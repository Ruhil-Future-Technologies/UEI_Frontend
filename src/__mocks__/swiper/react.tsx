import React from 'react';

export const Swiper: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => <div>{children}</div>;

export const SwiperSlide: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => <div>{children}</div>;
