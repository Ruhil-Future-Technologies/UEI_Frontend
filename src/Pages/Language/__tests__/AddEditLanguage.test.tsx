/* eslint-disable @typescript-eslint/no-explicit-any */
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import React from "react";
import AddEditLanguage from "../AddEditLanguage";
import NameContext from "../../Context/NameContext";
import {
  BrowserRouter,
  useLocation,
  useNavigate,
  useParams,
} from "react-router-dom";
import { contextValue } from "../../../MockStorage/mockstorage";


jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: jest.fn(),
  useNavigate: jest.fn(),
  useParams: jest.fn(),
}));

const mockedNavigate = jest.fn();

beforeEach(() => {
  (useLocation as jest.Mock).mockReturnValue({
    pathname: "localhost:3000/main/Language/add-Language",
  });
  (useParams as jest.Mock).mockReturnValue({ id: "12" }); // Mock 'id' consistently
  (useNavigate as jest.Mock).mockReturnValue(mockedNavigate); // Mock navigate function
});

afterEach(() => {
  jest.clearAllMocks();
});

describe("Add Edit Language Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  const renderComponent = () =>
    render(
      <NameContext.Provider value={contextValue}>
        <BrowserRouter>
          <AddEditLanguage />
        </BrowserRouter>
      </NameContext.Provider>
    );
  it("should render the add edit language component with all fields properly", () => {
    renderComponent();
    expect(screen.getByTestId("language_name")).toBeInTheDocument();
    expect(screen.getByTestId("language_file")).toBeInTheDocument();
    expect(screen.getByTestId("language_description")).toBeInTheDocument();
    expect(screen.getByTestId("submitBtn")).toBeInTheDocument();
  });

  it("should display validation errors for empty form submission", async () => {
    renderComponent();
    fireEvent.click(screen.getByTestId("submitBtn"));
    expect(
      await screen.findByText("Please enter Language name")
    ).toBeInTheDocument();
  });

  it("should display validation errors for invalid language_name", async () => {
    renderComponent();
    const input = screen.getByTestId("language_name");
    fireEvent.change(input, { target: { value: " " } }); // Add a space
    fireEvent.click(screen.getByTestId("submitBtn"));
    // Assert the validation error
    expect(
      await screen.findByText(
        "Please enter a valid Language name; whitespace is not allowed."
      )
    ).toBeInTheDocument();
  });
 
  it("should handle form submission successfully in add mode", async () => {
    renderComponent();
  
    // Fill in valid form data
    const languageNameInput = screen.getByTestId("language_name");
    const fileInput = screen.getByTestId("language_file");
    const descriptionInput = screen.getByTestId("language_description");
  
    // Simulate entering valid data for the language name, file, and description
    fireEvent.change(languageNameInput, { target: { value: "English" } });
    fireEvent.change(descriptionInput, { target: { value: "A widely spoken language." } });
  
    // Simulate file upload (mock file input)
    const mockFile = new File(["dummy content"], "dummy-image.png", { type: "image/png" });
    fireEvent.change(fileInput, { target: { files: [mockFile] } });
  
    // Submit the form
    fireEvent.click(screen.getByTestId("submitBtn"));
  
    // Assert that the button text is still "Save" because we're in add mode
    await screen.findByText("Update");
  
    // Check if handleSubmit has been triggered (mocked postData and other actions)
    expect(mockedNavigate).toHaveBeenCalledWith('/main/Language'); // Assuming successful navigation
  
    // If you are not sure about the exact success message, use waitFor to wait for a success message to appear
    await waitFor(() => {
      expect(screen.getByText(/language created successfully/i)).toBeInTheDocument();
    });
  
    // Verify that the form reset function was called to clear form data
    expect(screen.getByTestId("language_name")).toHaveValue(""); // Check if the field was reset
    expect(screen.getByTestId("language_description")).toHaveValue(""); // Check if description is reset
  });
  
  
   
});
