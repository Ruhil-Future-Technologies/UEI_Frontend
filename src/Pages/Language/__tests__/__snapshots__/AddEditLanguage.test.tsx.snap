// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Edit Language Component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="main-wrapper"
  >
    <div
      class="main-content"
    >
      <div
        class="card p-lg-3"
      >
        <div
          class="card-body"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 mb-3 css-2ulfj5-MuiTypography-root"
          >
            <div
              class="main_title"
            >
              Add Language
            </div>
          </h6>
          <form
            action="#"
            data-testid="form"
          >
            <div
              class="row"
            >
              <div
                class="col-md-4"
              >
                <div
                  class="form_field_wrapper"
                >
                  <div
                    class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                    field="[object Object]"
                    form="[object Object]"
                  >
                    <label
                      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                      data-shrink="false"
                      for=":r0:"
                      id=":r0:-label"
                    >
                      Language Name *
                    </label>
                    <div
                      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                    >
                      <input
                        aria-invalid="false"
                        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                        data-testid="language_name"
                        id=":r0:"
                        type="text"
                        value=""
                      />
                      <fieldset
                        aria-hidden="true"
                        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                      >
                        <legend
                          class="css-yjsfm1"
                        >
                          <span>
                            Language Name *
                          </span>
                        </legend>
                      </fieldset>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="col-md-4 mt-2"
              >
                <div
                  class="col"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
                  >
                    <h6
                      class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
                    >
                      Upload a Photo
                    </h6>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
                  >
                    <input
                      accept="image/*"
                      data-testid="language_file"
                      id="file-upload"
                      type="file"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="row mt-4"
            >
              <div
                class="col-md-4"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated text-secondary css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
                >
                  Description
                </label>
                <textarea
                  aria-label="empty textarea"
                  data-testid="language_description"
                  name="description"
                  placeholder="Enter your text here..."
                  style="width: 100%; font-size: 1rem;"
                />
              </div>
            </div>
            <div
              class="mt-3"
            >
              <button
                class="btn btn-primary mainbutton"
                data-testid="submitBtn"
                type="submit"
              >
                Save
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
